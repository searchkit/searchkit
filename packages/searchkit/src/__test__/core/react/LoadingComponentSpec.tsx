import * as React from "react";

import {
  SearchkitManager,
  LoadingComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    this.wrapper = mount(
      <LoadingComponent searchkit={this.searchkit}>
        <p>loading...</p>
      </LoadingComponent>
    )
  })

  it("display nothing when not loading", ()=> {
    expect(this.wrapper).toMatchSnapshot()
  })

  it("display children when loading", ()=> {
    this.searchkit.loading = true
    this.wrapper.instance().forceUpdate()
    expect(this.wrapper).toMatchSnapshot()
  })


})
