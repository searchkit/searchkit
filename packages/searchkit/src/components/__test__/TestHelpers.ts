export const fastClick = (el)=>{
  el.simulate("click")
}