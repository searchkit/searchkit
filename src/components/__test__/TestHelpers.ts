import ReactTestUtils = require('react-addons-test-utils')
import {html as beautifyHtml} from 'js-beautify'
import { renderToStaticMarkup } from 'react-dom/server'
import * as ReactDOM from "react-dom"
import {compact} from "lodash"
import {map} from "lodash"

export const hasClass = (inst, className)=> {
  if(ReactTestUtils.isDOMComponent(inst.node)) {
    return inst.hasClass(className)
  } else {
    try {
      let classes = ReactDOM.findDOMNode(inst.node).className
      return (' ' + classes + ' ').indexOf(' ' + className + ' ') > -1
    } catch (e){}
  }
  return false
}


export function jsxToHTML(Element){
  return renderToStaticMarkup(Element)
    .replace(/<input([^>]*)\/>/g, "<input$1>")
}

export function htmlClean(html) {
  return html
    .replace(/<!-- react-text: \d+ -->/g, '')
    .replace(/<!-- \/react-text -->/g, '')
    .replace(/<!-- react-empty: \d+ -->/g, '')
}

export const printPrettyHtml = (html)=> {
  html = beautifyHtml( html, {"indent_size":2} )
    .replace(/class=/g, "className=")
    .replace(/<input([^>]*)>/g, "<input$1/>")
    .replace(/readonly=""/g,"readOnly={true}")
    .replace(/font-size/g,"fontSize")
    .replace(/style="([^"]+)"+/g, (match, style)=> {
      let reactStyle = map(compact(style.split(";")), (keyvalue: string)=> {
        let [key, value] = keyvalue.split(":")
        return `${key}:"${value}"`
      }).join(",")
      return "style={{" + reactStyle + "}}"
    })
  console.log("\n"+ html)
}


export const fastClick = (el)=>{
  el.simulate("mouseDown", {button:0})
}
