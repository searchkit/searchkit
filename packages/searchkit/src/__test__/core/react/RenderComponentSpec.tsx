import * as React from "react";
import * as PropTypes from "prop-types";
import {mount} from "enzyme";
import {
  fastClick
} from "../../../components/__test__/TestHelpers"

import {Panel} from "../../../components"

import {
  renderComponent,
  RenderComponentType,
  RenderComponentPropType
} from "../../../"


describe("RenderComponent", ()=> {

  beforeEach(()=> {
    this.SubPanel = class SubPanel extends Panel {
    }
    this.SubPanel.defaultProps.title = "SubPanel"
    this.SubPanelElement = <Panel title="PanelElement"/>

    this.PanelReactClass = class PanelReactClass extends React.Component {
      static contextTypes = {
        color:PropTypes.string
      }
      render(){
        return (<Panel title={"PanelReactClass " + this.context.color} {...this.props}>
          {this.props.children}
        </Panel>)
      }
    }



    this.PanelFunction = (props, context)=> {
      return (
        <Panel title={"PanelFunction " + context.color}>
          {props.children}
        </Panel>
      )
    }
    this.PanelFunction.contextTypes = {
      color:PropTypes.string
    }

    class Provider extends React.Component<any, any>{
      static childContextTypes = {
        color: PropTypes.string
      }
      getChildContext(){
        return {color:"purple"}
      }
      render(){
        return this.props.children
      }
    }

    this.mount = (component, props={})=> {
      this.wrapper = mount(
        <Provider>
          {renderComponent(
            component, props,
            <p>content..</p>
          )}
        </Provider>
      )
    }

  })

  it("React.Component class", ()=> {
    this.mount(this.SubPanel)
    expect(this.wrapper).toMatchSnapshot()

  })

  it("React Element", ()=> {
    this.mount(this.SubPanelElement)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("React Class", ()=> {
    this.mount(this.PanelReactClass)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("Render function", ()=> {
    this.mount(this.PanelFunction)
    expect(this.wrapper).toMatchSnapshot()
  })

  it("Invalid component", ()=> {
    spyOn(console, "warn")
    try{
      this.mount(10)
    } catch (e){

    }
    expect(console.warn).toHaveBeenCalledWith(
      "Invalid component", 10
    )

  })

})
