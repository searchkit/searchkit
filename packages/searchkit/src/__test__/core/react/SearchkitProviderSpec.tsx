import * as React from "react";

import {
  SearchkitProvider,
  SearchkitManager,
  SearchkitComponent
} from "../../../"

import {mount} from "enzyme";

describe("SearchkitProvider", ()=> {

  beforeEach(()=> {
    this.searchkit = SearchkitManager.mock()
    class SomeComponent extends SearchkitComponent<any, any> {

      render(){
        return <h1>Hello</h1>
      }
    }
    this.wrapper = mount(
      <SearchkitProvider searchkit={this.searchkit}>
        <SomeComponent/>
      </SearchkitProvider>
    )
  })

  it("searchkit provider should work correctly", ()=> {
    expect(this.wrapper.html()).toBe("<h1>Hello</h1>")
    expect(this.wrapper.node.props.searchkit)
      .toBe(this.searchkit)
  })

  it("should call setupListeners()", ()=> {
    spyOn(this.searchkit, "setupListeners")
    expect(this.searchkit.setupListeners).not.toHaveBeenCalled()
    this.wrapper.node.componentWillMount()
    expect(this.searchkit.setupListeners).toHaveBeenCalled()
    this.searchkit.guidGenerator.counter = 10
    this.searchkit.unlistenHistory = jasmine.createSpy("unlisten")
    this.wrapper.node.componentWillUnmount()
    expect(this.searchkit.unlistenHistory).toHaveBeenCalled()
    expect(this.searchkit.guidGenerator.counter).toEqual(0)

  })


})
