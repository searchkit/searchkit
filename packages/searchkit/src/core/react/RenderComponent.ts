import * as React from 'react'
import * as PropTypes from "prop-types"

const omitBy = require("lodash/omitBy")
const isUndefined = require("lodash/isUndefined")

export type RenderFunction = (props?:any, children?:any) => Element
export type Element = React.ReactElement<any>
export type RenderComponentType<P> = React.ComponentClass<P> | React.ClassicComponentClass<P> | Element | RenderFunction | any

export const RenderComponentPropType = PropTypes.oneOfType([
  function(props:any, propName: string) {
     if(isUndefined(props[propName]) || (props[propName]["prototype"] instanceof React.Component)) {
       return null
     }
  },
  PropTypes.element,
  PropTypes.func,
])

export function renderComponent(component:RenderComponentType<any>, props={}, children=null): React.ReactElement<any> {
  let isReactComponent = (
    component["prototype"] instanceof React.Component ||
    (component["prototype"] && component["prototype"].isReactComponent) ||
    typeof component === 'function'
  )
  if (isReactComponent){
    return React.createElement(
      component as React.ComponentClass<any>,
      props, children
    )
  } else if (React.isValidElement(component)){
    return React.cloneElement(
      component as Element,
      omitBy(props, isUndefined), children
    );
  }
  console.warn("Invalid component", component)
  return null
}
