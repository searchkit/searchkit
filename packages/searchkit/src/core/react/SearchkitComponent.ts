import * as React from "react"
import * as PropTypes from "prop-types"
import {SearchkitManager} from "../SearchkitManager"
import {ImmutableQuery} from "../query"
import {Accessor} from "../accessors/Accessor"
import {Utils} from "../support"
const mapValues = require("lodash/mapValues")
import {block} from "./block"

export interface SearchkitComponentProps {
  mod?:string
  className?:string
  translations?:Object
  searchkit?:SearchkitManager
  key?:string
}

export class SearchkitComponent<P extends SearchkitComponentProps,S> extends React.Component<P,S> {
  _searchkit:SearchkitManager
  accessor:Accessor
  stateListenerUnsubscribe:Function
  translations:Object = {}
  unmounted = false

  static contextTypes: React.ValidationMap<any> = {
		searchkit:PropTypes.instanceOf(SearchkitManager)
	}

  static translationsPropType = (translations)=> {
    return PropTypes.shape(mapValues(translations, ()=> PropTypes.string))
  }

  static propTypes:any = {
    mod :PropTypes.string,
    className :PropTypes.string,
    translations: PropTypes.objectOf(
      PropTypes.string),
    searchkit:PropTypes.instanceOf(SearchkitManager)
  }

  constructor(props?){
    super(props)
    this.translate = this.translate.bind(this)
  }

  defineBEMBlocks() {
    return null;
  }

  defineAccessor():Accessor{
    return null
  }

  translate(key, interpolations?){
    let translation = (
      (this.searchkit.translate(key)) ||
      (this.props.translations && this.props.translations[key]) ||
      this.translations[key] || key)
    return Utils.translate(translation, interpolations)
  }

  get bemBlocks(): any {
    return mapValues(this.defineBEMBlocks(), (cssClass) => {
      return block(cssClass).el
    })
  }

  get searchkit(): SearchkitManager {
    return this._searchkit || (this._searchkit = this._getSearchkit());
  }

  _getSearchkit(): SearchkitManager {
    return this.props.searchkit || this.context["searchkit"];
  }

  set searchkit(value: SearchkitManager) {
    this._searchkit = value;
  }

  componentDidMount() {
    this._initAccessor();
    if (this.searchkit) {
      this.stateListenerUnsubscribe = this.searchkit.emitter.addListener(() => {
        if (!this.unmounted) {
          this.forceUpdate();
        }
      });
    }
    this.forceUpdate();
  }
  
  _initAccessor() {
    if (this.searchkit && !this.accessor) {
      this.accessor = this.defineAccessor();
      if (this.accessor) {
        this.accessor = this.searchkit.addAccessor(this.accessor);
        return true;
      }
    }
    if (!this.searchkit) {
      console.warn(
        "No searchkit found in props or context for " + this.constructor["name"]
      );
    }
    return false;
  }

  componentWillUnmount(){
    if(this.stateListenerUnsubscribe){
		  this.stateListenerUnsubscribe()
    }
    if(this.searchkit && this.accessor){
      this.searchkit.removeAccessor(this.accessor)
    }
    this.unmounted = true
	}

  getResults(){
    return this.searchkit.results
  }

  getHits(){
    return this.searchkit.getHits()
  }

  getHitsCount(){
    return this.searchkit.getHitsCount()
  }

  hasHits(){
    return this.searchkit.hasHits()
  }

  hasHitsChanged(){
    return this.searchkit.hasHitsChanged()
  }

  getQuery():ImmutableQuery {
    return this.searchkit.query
  }

  isInitialLoading(){
    return this.searchkit.initialLoading
  }

  isLoading(){
    return this.searchkit.loading
  }

  getError(){
    return this.searchkit.error
  }


}
