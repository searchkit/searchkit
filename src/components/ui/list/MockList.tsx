import * as React from "react";
import {assign} from "lodash"

export class MockList extends React.Component<any, any> {

  constructor(props){
    super(props)
    this.state = {
      items:[
        {key:"a", label:"A", doc_count:10},
        {key:"b", label:"B", doc_count:11, disabled:true},
        {key:"c", title:"C", doc_count:12},
        {key:"d", doc_count:15},
      ],
      docCount: 10+11+12+15,
      selectedItems:["a", "c"],
      toggleItem:jasmine.createSpy("toggleItem"),
      setItems:jasmine.createSpy("setItems"),
      translate(key){
        return key + " translated"
      },
      countFormatter:(count)=> "#"+count
    }
  }

  render(){
    return React.createElement(
      this.props.listComponent,
      assign({}, this.state, this.props)
    )
  }
}
