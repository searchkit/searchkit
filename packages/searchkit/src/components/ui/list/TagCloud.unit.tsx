import * as React from "react";
import {mount} from "enzyme";

import { TagCloud } from "./TagCloud"
import { MockList } from "./MockList"

import {fastClick} from "../../__test__/TestHelpers"

describe("TagCloud", ()=> {


  it("should render and behave correctly", ()=> {
    this.wrapper = mount(
      <MockList listComponent={TagCloud}/>
    )
    expect(this.wrapper).toMatchSnapshot()   

    this.wrapper.setProps({disabled:true})
    expect(this.wrapper.find(".sk-tag-cloud").hasClass("is-disabled")).toBe(true)

    expect(this.wrapper.instance().state.toggleItem).not.toHaveBeenCalled()
    fastClick(this.wrapper.find(".sk-tag-cloud__item").at(2))
    expect(this.wrapper.instance().state.toggleItem).toHaveBeenCalledWith("c")
  })

  it("should sort items", () => {
    let items = [
      { key: "d", doc_count: 15 },
      { key: "a", label: "a", doc_count: 10 },
      { key: "c", title: "C", doc_count: 12 },
      { key: "b", label: "B", doc_count: 11 },
    ]

    this.wrapper = mount(
      <MockList listComponent={TagCloud} items={items}/>
    )
    expect(this.wrapper).toMatchSnapshot()
  })

  it("mod + classname can be updated", () => {
    this.wrapper = mount(
      <MockList listComponent={TagCloud} mod="sk-other-class" className="my-custom-class"/>
    )

    expect(this.wrapper.find(".sk-other-class").hasClass("my-custom-class")).toBe(true)
  })

  it("show count", () => {
    this.wrapper = mount(
      <MockList listComponent={TagCloud} showCount={true}/>
    )

    expect(this.wrapper.find(".sk-tag-cloud__item").at(0).text())
      .toBe("A translated#10") // Count is appended in a span and styled with CSS
  })

})
