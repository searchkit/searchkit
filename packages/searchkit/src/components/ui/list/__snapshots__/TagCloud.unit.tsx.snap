// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagCloud should render and behave correctly 1`] = `
<div class="sk-tag-cloud">
  <div class="sk-tag-cloud-option sk-tag-cloud__item is-active" style="font-size: 1em;" data-qa="option" data-key="a">
    <div data-qa="label" class="sk-tag-cloud-option__text">A translated</div>
  </div>
  <div class="sk-tag-cloud-option sk-tag-cloud__item is-disabled" style="font-size: 1.1em;" data-qa="option" data-key="b">
    <div data-qa="label" class="sk-tag-cloud-option__text">B translated</div>
  </div>
  <div class="sk-tag-cloud-option sk-tag-cloud__item is-active" style="font-size: 1.2em;" data-qa="option" data-key="c">
    <div data-qa="label" class="sk-tag-cloud-option__text">C translated</div>
  </div>
  <div class="sk-tag-cloud-option sk-tag-cloud__item" style="font-size: 1.5em;" data-qa="option" data-key="d">
    <div data-qa="label" class="sk-tag-cloud-option__text">d translated</div>
  </div>
</div>
`;

exports[`TagCloud should sort items 1`] = `
<div class="sk-tag-cloud">
  <div class="sk-tag-cloud-option sk-tag-cloud__item is-active" style="font-size: 1em;" data-qa="option" data-key="a">
    <div data-qa="label" class="sk-tag-cloud-option__text">a translated</div>
  </div>
  <div class="sk-tag-cloud-option sk-tag-cloud__item" style="font-size: 1.1em;" data-qa="option" data-key="b">
    <div data-qa="label" class="sk-tag-cloud-option__text">B translated</div>
  </div>
  <div class="sk-tag-cloud-option sk-tag-cloud__item is-active" style="font-size: 1.2em;" data-qa="option" data-key="c">
    <div data-qa="label" class="sk-tag-cloud-option__text">C translated</div>
  </div>
  <div class="sk-tag-cloud-option sk-tag-cloud__item" style="font-size: 1.5em;" data-qa="option" data-key="d">
    <div data-qa="label" class="sk-tag-cloud-option__text">d translated</div>
  </div>
</div>
`;
